# Apriori Algorithm

## Main Task
> Identify the best deals to maximize the chance that customers will get the deals. (Buy the product X and get the product Y for free, this can optimize the sales and the profit).

### Data Understanding  

**1.0. What is the domain area of the dataset?**  
The dataset *Market_Basket_Optimisation.csv* contains information about people who have visited a shopping centrum. 

**2.0. Which data format?**  
The dataset is in *csv* format!  

**2.1. Do the files have headers or another file describing the data?**  
The files does have headers that describes the data! Each column has a name that describes the data it contains!  

**2.2. Are the data values separated by commas, semicolon, or tabs?**  
The data values are separated by commas!  
Example: 
*burgers,meatballs,eggs*

**3.0 How many features and how many observations does the dataset have?**  
The dataset has:  
* 5 features or columns!
* 200 observations or rows!  

**4.0 Does it contain numerical features? How many?**  
No

**5.0. Does it contain categorical features?  How many?**  
No, but they can be categorized.

In [3]:
# In case apyori module is not installed on your machine!
# !pip install apyori

In [4]:
# Importing necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from apyori import apriori

### Data Preprocessing

In [5]:
dataset = pd.read_csv("../Dataset/Market_Basket_Optimisation.csv")

In [6]:
dataset.head()

shrimp    almonds     avocado    vegetables mix green grapes  \
0         burgers  meatballs        eggs               NaN          NaN   
1         chutney        NaN         NaN               NaN          NaN   
2          turkey    avocado         NaN               NaN          NaN   
3   mineral water       milk  energy bar  whole wheat rice    green tea   
4  low fat yogurt        NaN         NaN               NaN          NaN   

  whole weat flour yams cottage cheese energy drink tomato juice  \
0              NaN  NaN            NaN          NaN          NaN   
1              NaN  NaN            NaN          NaN          NaN   
2              NaN  NaN            NaN          NaN          NaN   
3              NaN  NaN            NaN          NaN          NaN   
4              NaN  NaN            NaN          NaN          NaN   

  low fat yogurt green tea honey salad mineral water salmon antioxydant juice  \
0            NaN       NaN   NaN   NaN           NaN    NaN               NaN   
1            NaN       NaN   NaN   NaN           NaN    NaN               NaN   
2            NaN       NaN   NaN   NaN           NaN    NaN               NaN   
3            NaN       NaN   NaN   NaN           NaN    NaN               NaN   
4            NaN       NaN   NaN   NaN           NaN    NaN               NaN   

  frozen smoothie spinach  olive oil  
0             NaN     NaN        NaN  
1             NaN     NaN        NaN  
2             NaN     NaN        NaN  
3             NaN     NaN        NaN  
4             NaN     NaN        NaN

In [7]:
print(f"Number of features in the dataset is {dataset.shape[1]} and the number of observations/rows in the dataset is {dataset.shape[0]}")

Number of features in the dataset is 20 and the number of observations/rows in the dataset is 7500


In [8]:
# for using aprori we need to convert data in list format..
# transactions = [['apple','almonds'],['apple'],['banana','apple']]....

transactions = []
for i in range(0, len(dataset)):
    temp = []
    for j in range(0, 20):
        tempStr = str(dataset.values[i, j]) #dataset.values[rows, columns]
        if(tempStr != 'nan'):
            temp.append(tempStr)
    transactions.append(temp)

In [9]:
transactions[0]

['burgers', 'meatballs', 'eggs']

### Training the Apriori Model on the dataset

**min_support:** This is the minimum support value for an itemset to be considered frequent. If the support of an itemset is lower than this value, it’s ignored. The support of an itemset is the proportion of transactions in the dataset that contain the itemset. 

**min_confidence:** This is the minimum confidence value for a rule to be considered significant. If the confidence of a rule is lower than this value, it’s ignored. The confidence of a rule is the proportion of transactions that contain the antecedent of the rule, which also contain the consequent.

**min_lift:** This is the minimum lift value for a rule to be considered significant. If the lift of a rule is lower than this value, it’s ignored. The lift of a rule is the ratio of the observed support to that expected if the antecedent and the consequent were independent.

**min_length** and **max_length:** These parameters specify the minimum and maximum length of the itemsets/rules to be considered.  

> The appropriate value for hypoerparameters depends on your domain and how strict you want to be with your rules.

In [10]:
rules = apriori(transactions=transactions,
                min_support = 0.003, min_confidence = 0.2,
                min_lift = 3, min_length = 2, max_length = 2)

### Visualising the results

In [11]:
results = list(rules)
results

[RelationRecord(items=frozenset({'light cream', 'chicken'}), support=0.004533333333333334, ordered_statistics=[OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.2905982905982906, lift=4.843304843304844)]),
 RelationRecord(items=frozenset({'mushroom cream sauce', 'escalope'}), support=0.005733333333333333, ordered_statistics=[OrderedStatistic(items_base=frozenset({'mushroom cream sauce'}), items_add=frozenset({'escalope'}), confidence=0.30069930069930073, lift=3.7903273197390845)]),
 RelationRecord(items=frozenset({'pasta', 'escalope'}), support=0.005866666666666667, ordered_statistics=[OrderedStatistic(items_base=frozenset({'pasta'}), items_add=frozenset({'escalope'}), confidence=0.37288135593220345, lift=4.700185158809287)]),
 RelationRecord(items=frozenset({'fromage blanc', 'honey'}), support=0.0033333333333333335, ordered_statistics=[OrderedStatistic(items_base=frozenset({'fromage blanc'}), items_add=frozenset({'honey'}), confiden

### Putthing the results into a Pandas DataFrame

In [12]:
def inspect(results):
    lhs         = [tuple(result[2][0][0])[0] for result in results]
    rhs         = [tuple(result[2][0][1])[0] for result in results]
    supports    = [result[1] for result in results]
    confidences = [result[2][0][2] for result in results]
    lifts       = [result[2][0][3] for result in results]
    return list(zip(lhs, rhs, supports, confidences, lifts))
resultsinDataFrame = pd.DataFrame(inspect(results), columns = ['Left Hand Side', 'Right Hand Side', 'Support', 'Confidence', 'Lift'])

### The results (non-sorted)

In [13]:
resultsinDataFrame

Left Hand Side Right Hand Side   Support  Confidence      Lift
0           light cream         chicken  0.004533    0.290598  4.843305
1  mushroom cream sauce        escalope  0.005733    0.300699  3.790327
2                 pasta        escalope  0.005867    0.372881  4.700185
3         fromage blanc           honey  0.003333    0.245098  5.178128
4         herb & pepper     ground beef  0.016000    0.323450  3.291555
5          tomato sauce     ground beef  0.005333    0.377358  3.840147
6           light cream       olive oil  0.003200    0.205128  3.120612
7     whole wheat pasta       olive oil  0.008000    0.271493  4.130221
8                 pasta          shrimp  0.005067    0.322034  4.514494

### The results (sorted)

In [14]:
resultsinDataFrame.nlargest(n = 10, columns='Lift')

Left Hand Side Right Hand Side   Support  Confidence      Lift
3         fromage blanc           honey  0.003333    0.245098  5.178128
0           light cream         chicken  0.004533    0.290598  4.843305
2                 pasta        escalope  0.005867    0.372881  4.700185
8                 pasta          shrimp  0.005067    0.322034  4.514494
7     whole wheat pasta       olive oil  0.008000    0.271493  4.130221
5          tomato sauce     ground beef  0.005333    0.377358  3.840147
1  mushroom cream sauce        escalope  0.005733    0.300699  3.790327
4         herb & pepper     ground beef  0.016000    0.323450  3.291555
6           light cream       olive oil  0.003200    0.205128  3.120612

## Notes
The conclusion you can draw from these results depends on what you’re looking for in your association rules.

The Lift value indicates the likelihood of the items in the right-hand side being purchased when the items in the left-hand side are purchased.  
A Lift value greater than 1 indicates that the items are likely to be bought together, and the higher the Lift value, the stronger the association.

So, if you’re looking for the strongest associations, then yes, the rule with the highest Lift (in this case, fromage blanc -> honey with a Lift of 5.178128) would be considered the “best”.

However, it’s important to also consider the Support and Confidence values.

* Support is the proportion of transactions in the dataset that contain the itemset. A low support might indicate that the rule doesn’t apply to many transactions, even if the Lift is high.  

* Confidence is the proportion of transactions with the left-hand side item(s) that also contain the right-hand side item(s). A low confidence might indicate that the rule isn’t reliable, even if the Lift is high.